### PR TITLE
Feature/socket improvement

### PR DESCRIPTION
**Pull request by eveshi**
1. Add `JWT token` for connection.
2. Change broadcast function to `target push`.
3. Move `token` from local storage to cookie.
4. Update `onlineUser` data structure.

### CODE REVIEW

File: server/bin/www [hunk 1]
```diff
@@ -31,22 +35,50 @@ const server = http.createServer(app);
 
 const io = require("socket.io")(server);
 
+// add token verification
+io.use((socket, next) => {
```
Comment (kuperavv) [inline server/bin/www:39]: Great!

File: server/bin/www [hunk 2]
```diff
@@ -31,22 +35,50 @@ const server = http.createServer(app);
...
       onlineUsers.push(id);
     }
+    socket.join(id);
     // send the user who just went online to everyone else who is already online
     socket.broadcast.emit("add-online-user", id);
   });
 
   socket.on("new-message", (data) => {
-    socket.broadcast.emit("new-message", {
+    socket.to(data.recipientId).to(data.senderId).emit("new-message", {
```
Comment (kuperavv) [inline server/bin/www:68]: Yes, this def was a security issue.

File: server/bin/www [hunk 3]
```diff
@@ -31,22 +35,50 @@ const server = http.createServer(app);
...
+    if (user){
+      return next()
+    }
+    });
+  }
+});
+
 io.on("connection", (socket) => {
   socket.on("go-online", (id) => {
     if (!onlineUsers.includes(id)) {
```
Comment (kuperavv) [inline server/bin/www:59]: This includes is inefficient.  Perhaps a data structure with constant look-up time would be better?

**Review by kuperavv — COMMENTED** [review]
In general please try to use separate branches for each ticket.  It makes providing more challenging.

I left a comment to address regarding the onlineUsers data structure.  There's also a subtle issue with your cookie branch.  When you logout, and then refresh the page, your session is persisted and you are logged back in.  How are you clearing the httpOnly cookie?

This likely also masked an issue with your socket implementation.  When the front end socket initially attempts to connect with an unauthenticated user, the server rightly declines the socket connection, but the socket never attempts to connect again after the user does authenticate.

**Review by kuperavv — COMMENTED** [review]
If you can, see if you can figure out how to defer the socket.io connection until after the user authenticates to avoid continuously polling the backend for an unauthenticated user.  This could a drain on server capacity if you have a large numbers of users sitting on your signup and login pages.

File: client/src/store/utils/thunkCreators.js [hunk 4]
```diff
@@ -36,8 +32,9 @@ export const fetchUser = () => async (dispatch) => {
 export const register = (credentials) => async (dispatch) => {
   try {
     const { data } = await axios.post("/auth/register", credentials);
-    await localStorage.setItem("messenger-token", data.token);
+    // await localStorage.setItem("messenger-token", data.token);
```
Comment (ThinkTink) [inline client/src/store/utils/thunkCreators.js:35]: let's remove commented out code completely, to keep our code base clean

File: server/routes/auth/index.js [hunk 5]
```diff
@@ -26,9 +26,9 @@ router.post("/register", async (req, res, next) => {
       process.env.SESSION_SECRET,
       { expiresIn: 86400 }
     );
+    res.cookie("token", token, { httpOnly: true });
```
Comment (ThinkTink) [inline server/routes/auth/index.js:29]: We should set an expiration date for the cookie as well, so it's not cleared when the browser is closed

**Review by ThinkTink — APPROVED** [review]
:+1:

**Pull request merged**